jetty-server/src/test/java/org/eclipse/jetty/server/ErrorHandlerTest.java

```diff
@@ -302,29 +385,150 @@ public void testBadMessage() throws Exception
+        assertThat("Response Content-Type", response.get(HttpHeader.CONTENT_TYPE), containsString("text/html;charset=ISO-8859-1"));
+
+        assertContent(response);
+    }
+
+    @ParameterizedTest
+    @ValueSource(strings = {
+        "/jsonmessage/",
+        "/xmlmessage/",
+        "/htmlmessage/",
+        "/utf8message/",
+    })
+    public void testComplexCauseMessageNoAcceptHeader(String path) throws Exception
+    {
+        String rawResponse = connector.getResponse(
+            "GET " + path + " HTTP/1.1\r\n" +
+                "Host: Localhost\r\n" +
+                "\r\n");
+
+        HttpTester.Response response = HttpTester.parseResponse(rawResponse);
+
+        assertThat("Response status code", response.getStatus(), is(500));
+        assertThat("Response Content-Length", response.getField(HttpHeader.CONTENT_LENGTH).getIntValue(), greaterThan(0));
+        assertThat("Response Content-Type", response.get(HttpHeader.CONTENT_TYPE), containsString("text/html;charset=ISO-8859-1"));
+
+        String content = assertContent(response);
+
+        if (path.startsWith("/utf8"))
+        {
+            // we are Not expecting UTF-8 output, look for mangled ISO-8859-1 version
+            assertThat("content", content, containsString("Euro is &amp;euro; and ? and %E2%82%AC"));
+        }
+    }
+
+    @ParameterizedTest
+    @ValueSource(strings = {
+        "/jsonmessage/",
+        "/xmlmessage/",
+        "/htmlmessage/",
+        "/utf8message/",
+    })
+    public void testComplexCauseMessageAcceptUtf8Header(String path) throws Exception
+    {
+        String rawResponse = connector.getResponse(
+            "GET " + path + " HTTP/1.1\r\n" +
+                "Host: Localhost\r\n" +
+                "Accept: text/html\r\n" +
+                "Accept-Charset: utf-8\r\n" +
+                "\r\n");
+
+        HttpTester.Response response = HttpTester.parseResponse(rawResponse);
+
+        assertThat("Response status code", response.getStatus(), is(500));
+        assertThat("Response Content-Length", response.getField(HttpHeader.CONTENT_LENGTH).getIntValue(), greaterThan(0));
+        assertThat("Response Content-Type", response.get(HttpHeader.CONTENT_TYPE), containsString("text/html;charset=UTF-8"));
+
+        String content = assertContent(response);
+
+        if (path.startsWith("/utf8"))
+        {
+            // we are Not expecting UTF-8 output, look for mangled ISO-8859-1 version
+            assertThat("content", content, containsString("Euro is &amp;euro; and \u20AC and %E2%82%AC"));
+        }
+    }
+
+    private String assertContent(HttpTester.Response response)
+    {
+        String contentType = response.get(HttpHeader.CONTENT_TYPE);
+        String content = response.getContent();
+
+        if (contentType.contains("text/html"))
+        {
+            assertThat(content, not(containsString("<script>")));
```

Comment:
The usefulness of such an attack isn't as obvious as script tags in HTML, but we should assure that (potentially malformed) JSON or XML messages in exceptions are properly escaped as well. Eg. new Exception("\"}{\"I am injected JSON\":\"evil\"}{\"void\":\"") or new Exception("</something><injectedxml>evil</injectedxml><something>"), or whatever could break out of the JSON/XML structure and create or modify elements.

Reply:
That's been addressed.
The ErrorHandler only produces `text/html`, `text/json`, or `text/plain` output.

For `text/json` all content from exceptions (the stacktrace, messages, causes, and supressed) are filtered via [`StringUtil.sanitizeXmlString()`](https://github.com/eclipse/jetty.project/blob/ba1fe2826d49c52f82b49bb32737a361519c0479/jetty-util/src/main/java/org/eclipse/jetty/util/StringUtil.java#L1109-L1171)

For `text/json` all values are filtered via the same `StringUtil.sanitizeXmlString()`.

For `text/plain` they are left alone.

Reply:
I'll add another test for the JSON example you provided.

Reply:
Yeah, it's fine. 
See commit 217602e97f2b9de41e042620148f910a82626cd0

The generated `text/json` looks like this ...

``` json
{
"cause1":"java.lang.RuntimeException: &quot;}, &quot;glossary&quot;: {\n &quot;title&quot;: &quot;example&quot;\n }\n {&quot;",
"cause0":"javax.servlet.ServletException: java.lang.RuntimeException: &quot;}, &quot;glossary&quot;: {\n &quot;title&quot;: &quot;example&quot;\n }\n {&quot;",
"message":"javax.servlet.ServletException: java.lang.RuntimeException: &quot;}, &quot;glossary&quot;: {\n &quot;title&quot;: &quot;example&quot;\n }\n {&quot;",
"url":"/jsonmessage/",
"status":"500"
}
```